### PR TITLE
feat: add evm proxy history

### PR DESCRIPTION
**Pull request by hoangtv49**
## Summary by CodeRabbit

- **New Features**
	- Introduced a new database table to track historical data for Ethereum Virtual Machine (EVM) proxy contracts.
	- Added functionality for handling and tracking proxy contracts, including support for different proxy types and ERC1967 events.
	- Launched a new service for crawling Ethereum blockchain events related to proxy contracts, capable of tracking changes in admin and implementation addresses.

- **Refactor**
	- Updated contract type detection logic and simplified proxy detection methods.
	- Consolidated contract proxy detection into a new `ContractHelper` class for improved efficiency.

- **Chores**
	- Added and updated constants and configurations related to proxy contracts handling.

### CODE REVIEW

**Comment by hoangtv49** [comment]
@coderabbitai review

File: migrations/20240320080140_create-evm-proxy-history.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
+  await knex.schema.createTable('evm_proxy_history', (table) => {
+    table.increments();
+    table.string('proxy_contract').index().notNullable();
+    table.string('implementation_contract');
```
Comment (peara) [inline migrations/20240320080140_create-evm-proxy-history.ts:7]: index luôn

File: migrations/20240320080140_create-evm-proxy-history.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    table.string('admin');
+    table.integer('height');
+    table.string('hash');
+    table.unique([
```
Comment (peara) [inline migrations/20240320080140_create-evm-proxy-history.ts:11]: cái này to thế, có cần nó unique ko?

File: migrations/20240320080140_create-evm-proxy-history.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
+import { Knex } from 'knex';
+
+export async function up(knex: Knex): Promise<void> {
+  await knex.schema.createTable('evm_proxy_history', (table) => {
+    table.increments();
+    table.string('proxy_contract').index().notNullable();
+    table.string('implementation_contract');
+    table.string('admin');
+    table.integer('height');
+    table.string('hash');
```
Comment (peara) [inline migrations/20240320080140_create-evm-proxy-history.ts:10]: hash này là tx hash hay code hash

File: src/services/evm/crawl_proxy_evm.service.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+      `Crawl proxy contract from block ${startBlock} to block ${endBlock}`
+    );
+    const evmEvents = await EvmEvent.query()
+      .select('*')
+      .where('block_height', '>', startBlock)
+      .andWhere('block_height', '<=', endBlock);
+    const proxyContractDb = await EVMSmartContract.query().where(
+      'type',
+      'like',
+      'PROXY_%'
```
Comment (peara) [inline src/services/evm/crawl_proxy_evm.service.ts:57]: chắc là vẫn chạy, nhưng dùng whereIn đi, có mấy cái đâu

File: src/services/evm/crawl_proxy_evm.service.ts [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+          )?.SLOT
+        );
+      }
+
+      newProxyHistories.push(
+        EvmHistoryProxy.fromJson({
+          proxy_contract: _.toLower(evmEvent.address),
+          admin: _.toLower(adminAddress as string),
+          implementation_contract: _.toLower(implementationAddress as string),
+          height: evmEvent.block_height,
```
Comment (peara) [inline src/services/evm/crawl_proxy_evm.service.ts:128]: chỉ thêm nếu bắt được event upgraded

**Pull request merged**